src/action/finder/request.rs

```diff
@@ -0,0 +1,1405 @@
+        let process_non_regular_files = proto.process_non_regular_files();
+        let xdev_mode = parse_enum(proto.xdev)?;
+        let path_type = parse_enum(proto.pathtype)?;
+
+        let mut conditions = vec![];
+        for proto_condition in proto.conditions {
+            conditions.extend(get_conditions(proto_condition)?);
+        }
+
+        let action = match proto.action {
+            Some(action) => Action::try_from(action)?,
+            None => {
+                return Err(ParseError::malformed(
+                    "File Finder request does not contain action definition.",
+                ));
+            }
+        };
+
+        Ok(Request {
+            paths: proto.paths,
+            path_type,
+            action,
+            conditions,
+            follow_links,
+            process_non_regular_files,
+            xdev_mode,
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::action::Request as _;
+
+    #[test]
+    fn test_empty_request() {
+        let request = Request::from_proto(FileFinderArgs {
+            action: Some(FileFinderAction {
+                action_type: Some(ActionType::Stat as i32),
+                ..Default::default()
+            }),
+            ..Default::default()
+        })
+        .unwrap();
+
+        assert!(request.paths.is_empty());
+        assert!(request.conditions.is_empty());
+        assert_eq!(request.path_type, PathType::Os);
+        assert_eq!(request.process_non_regular_files, false);
+        assert_eq!(request.follow_links, false);
+        assert_eq!(request.xdev_mode, XDevMode::Local);
+    }
+
+    #[test]
+    fn test_basic_root_parameters() {
+        let request = Request::from_proto(FileFinderArgs {
+            paths: vec!["abc".to_string(), "cba".to_string()],
+            pathtype: Some(PathType::Registry as i32),
+            process_non_regular_files: Some(true),
+            follow_links: Some(true),
+            xdev: Some(rrg_proto::file_finder_args::XDev::Always as i32),
+            action: Some(FileFinderAction {
+                action_type: Some(ActionType::Stat as i32),
+                ..Default::default()
+            }),
+            ..Default::default()
+        })
+        .unwrap();
+
+        assert_eq!(request.paths, vec!["abc".to_string(), "cba".to_string()]);
+        assert_eq!(request.path_type, PathType::Registry);
+        assert_eq!(request.process_non_regular_files, true);
+        assert_eq!(request.follow_links, true);
+        assert_eq!(request.xdev_mode, XDevMode::Always);
+    }
+
+    #[test]
+    fn test_default_stats_action() {
+        let request = Request::from_proto(FileFinderArgs {
+            action: Some(FileFinderAction {
+                action_type: Some(ActionType::Stat as i32),
+                ..Default::default()
+            }),
+            ..Default::default()
+        })
+        .unwrap();
+
+        match request.action {
+            Action::Stat(options) => {
```

Comment:
Prefer to use `if let` to reduce nesting (here and below).

Reply:
Done!